### ISSUE TITLE
sysctl template

### ISSUE THREAD

**Issue by truzzon** [ComplianceAsCode/content#10203]
#### Share the context

I want to check if sysctl is configured properly:

I have a default config provided by the vendor (```/usr/lib/sysctl.d/...conf```). The values get overwritten by parameters saved in ```/etc/sysctl.d/*.conf```

Currently I get failed test even though the variables that are actually loaded are correct.

#### Description of problem:

The problem begins if e.g. the parameter ```fs.suid_dumpable``` is configured in ```/usr/lib/sysctl.d/``` and then overwritten by ```/etc/sysctl.d/custom.conf```.

Since both files are evaluated this causes (in my opinion) a false positive in the result, because in the end the value, that was loaded last is actually used. 

There is no verification what is actually active in memory.

#### Proposed change:

I would propose, that in addition to the file checks, the values that are in memory (```sysctl -a```) also get checked because this finally decides what is active on the system and what not. And if the parameter in memory are correct, the check should pass.

And I already talked to my customer: Simply deleting the files under ```/usr/lib/sysctl.d``` is not an option. Deleting is how I interpreted the tests of the current template.

**Comment by Xeicker** [comment]
I don't think this would be implemented. As this represent to handle any kind of precedence in how files are processed. The oval content does takes into account runtime settings, but also applies an and operation with configurations found in files.

Instead of deleting files, commenting the specific configurations which is causing the issue could be an option. 

Btw, you should never take tests as a way of configuring your system, those are intended only to validate the behavior of OVAL/remediation, and aren't made thinking on an usable system.

**Comment by antiphase** [comment]
I'm also experiencing a similar problem, and when systemd is used to manage sysctls, simply grepping config files for uncommented values is wholly inadequate.

Not only is there the read ordering of directories followed by lexical file naming to consider when processing the directives, systemd-sysctl also supports masking of entire files using symlinks to /dev/null (`ln -s /dev/null /etc/sysctl.d/50-default.conf` prevents `/usr/lib/sysctl.d/50-default.conf` from being processed) as well as wildcards; e.g. setting `net.ipv4.conf.*.rp_filter` sets `net.ipv4.conf.all.rp_filter` as well as updating any existing interfaces. There are also lines prefixed with `-` which prevent values being overridden later.

As usual, I don't think anyone actually asked for these "improvements", but here we are.

The only solutions are reimplementing the same logic here, which seems like a complex and brittle option, or simply reading the live values when auditing.

I would also second that commenting lines in vendor-distributed files is an antipattern, as package updates tend to undo such changes, and overrides as documented are the only acceptable means of persistent remediation.

**Comment by MitchDrage** [comment]
Here's a specific example, building on what others have said:

The requirement in alma9 CIS level 1 is `net.ipv4.conf.default.rp_filter = 1`, where `/usr/lib/sysctl.d/50-default.conf` sets the value to `2`, while our CIS-compliant configuration correctly overrides it to `1` via a drop-in in `/etc/sysctl.d/`.

The OVAL test results show:
- All files in `/etc/sysctl.d/` correctly show `rp_filter = 1` - evaluated as `true`
- `/usr/lib/sysctl.d/50-default.conf` shows `rp_filter = 2` - evaluated as `false`
- The runtime check (`sysctl net.ipv4.conf.default.rp_filter`) returns `1` - evaluated as `true`
- The overall rule still fails due to the `and` operation with the file check

This is the drop-in override pattern that systemd explicitly documents and recommends - see systemd issue [#32563](https://github.com/systemd/systemd/issues/32563) where this exact scenario was raised and closed as "not a bug" because the drop-in mechanism is the correct solution. The systemd project considers `/usr/lib/sysctl.d/50-default.conf` shipping `rp_filter = 2` intentional, and the correct override mechanism is exactly what we're using.

The suggestion from @Xeicker to comment out lines in vendor-distributed files is not viable - as @antiphase notes, this is an antipattern that will be undone by package updates, and is explicitly called out as wrong in the systemd documentation.

The runtime value is correct. The drop-in override is correct. The OVAL test is producing a false positive because it cannot model sysctl.d load-order precedence.

I'd support the proposal to treat a passing runtime check as sufficient. If needed, this could be limited to when a conflicting value exists only in a package-owned file under `/usr/lib/sysctl.d/`, given that user-space overrides in `/etc/sysctl.d/` are present and correct.

Edit:
It looks like for my specific issue might be related to the fact that AlmaLinux isn't in this list (in https://github.com/ComplianceAsCode/content/blob/master/shared/templates/sysctl/oval.template):
`      <object_reference>object_static_etc_sysctls_{{{ rule_id }}}</object_reference>
{{% if product not in [ "ol7", "ol8", "ol9", "rhcos4", "rhel8", "rhel9", "rhel10", "ubuntu2204", "ubuntu2404"] %}}
      <object_reference>object_static_lib_sysctld_{{{ rule_id }}}</object_reference>
{{% endif %}}`
